tests/PersonDialogTest.java

```diff
@@ -0,0 +1,204 @@
+    @AfterEach
+    void tearDown() {
+        // Close assertJ window gui
+        window.cleanUp();
+    }
+
+    @AfterAll
+    public static void clean() {
+        // Re-enable program to close after testing completes
+        NoExitSecurityManagerInstaller.installNoExitSecurityManager().uninstall();
+    }
+   
+    /**
+     * This tests that a person can be added.
+     */
+
+    @Test
+    void addPersonValid() {
+        window.button("add").click();
+        DialogFixture dialog = window.dialog();
+
+        // Type 'Muffin','Man','1234 Dreary Lane','Cookie City','WA','66666', and '1234567890'
+        // into the respective boxes
+        dialog.textBox("firstName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_M).releaseKey(VK_SHIFT)
+                .pressAndReleaseKeys(VK_U, VK_F, VK_F, VK_I, VK_N);
+        dialog.textBox("lastName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_M).releaseKey(VK_SHIFT)
+                .pressAndReleaseKeys(VK_A, VK_N);
+        dialog.textBox("address").pressAndReleaseKeys(VK_1, VK_2, VK_3, VK_4, VK_SPACE).pressKey(VK_SHIFT)
+                .pressAndReleaseKeys(VK_D).releaseKey(VK_SHIFT)
+                .pressAndReleaseKeys(VK_R, VK_E, VK_A, VK_R, VK_Y, VK_SPACE).pressKey(VK_SHIFT)
+                .pressAndReleaseKeys(VK_L).releaseKey(VK_SHIFT).pressAndReleaseKeys(VK_A, VK_N, VK_E);
+        dialog.textBox("city").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_C).releaseKey(VK_SHIFT)
+                .pressAndReleaseKeys(VK_O, VK_O, VK_K, VK_I, VK_E, VK_SPACE).pressKey(VK_SHIFT).pressAndReleaseKeys(VK_C)
+                .releaseKey(VK_SHIFT).pressAndReleaseKeys(VK_I, VK_T, VK_Y);
+        dialog.textBox("state").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_W, VK_A).releaseKey(VK_SHIFT);
+        dialog.textBox("zip").pressAndReleaseKeys(VK_6, VK_6, VK_6, VK_6, VK_6);
+        dialog.textBox("phone").pressAndReleaseKeys(VK_1, VK_2, VK_3, VK_4, VK_5, VK_6, VK_7, VK_8, VK_9, VK_0);
+
+        // Click 'OK'
+        dialog.button(JButtonMatcher.withText("OK")).click();
+
+        // Test person is added
+        window.table().requireRowCount(1);
+
+    }
+
+    /**
+     * This tests that a person cannot be added with improper characters.
+     */
+
+    @Test
+    void addPersonInvalid() {
+        // Test when user exits with OK
+        window.button("add").click();
+        DialogFixture dialog = window.dialog();
+
+        dialog.textBox("firstName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_1).releaseKey(VK_SHIFT)
+            .pressAndReleaseKeys(VK_2, VK_2, VK_4, VK_5, VK_6);
+        
+        dialog.textBox("lastName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_M).releaseKey(VK_SHIFT)
+            .pressAndReleaseKeys(VK_9, VK_7);
+        
+        // Click 'OK'
+        dialog.button(JButtonMatcher.withText("OK")).click();
+
+        // Test person is not added
+        window.table().requireRowCount(0);
+
+    }
+
+
+    @Test
+    void addPersonNullFirst() {
+        // Test when user exits with OK
```

Comment:
null and " " are not the same thing. This test could be renamed to addPersonEmptyFirstName()

Reply:
This test and the other seem fine, but you need to remove the dependency on the AddressBookGUI for verifying whether the "person" exists or not. If you "mock" the class, you should be able to check the mock to see if the person exists. If that makes sense.